polyA/calculate_score.py

```diff
@@ -70,5 +74,81 @@ def calculate_score(
                 chunk_score += gap_init
         else:
             chunk_score += sub_matrix[seq1_char + seq2_char]
-
+            if char_complexity_adjustment:
+                chunk_score += char_complexity_adjustment[seq2_char]
     return chunk_score
+
+
+def calc_target_char_counts(
+    query_seq: str,
+    target_seq: str,
+):
+    query_char_counts = {"A": 0, "C": 0, "G": 0, "T": 0}
+    other_chars = set()
+    total_chars = 0
+    for q_base, t_base in zip(query_seq, target_seq):
+        if t_base not in ["A", "G", "T", "C"]:
+            other_chars.add(t_base)
+            continue
+        if q_base == "-":
+            continue
+        query_char_counts[t_base] += 1
+        total_chars += 1
+    return query_char_counts, total_chars, other_chars
+
+
+def calculate_complexity_adjusted_score(
+    char_background_freqs: Optional[Dict[str, float]],
+    query_seq: str,
+    target_seq: str,
+    lamb: float,
+):
+    char_complexity_adjustments: Dict[str, float] = {}
+    if char_background_freqs is None:
+        # set the complexity adjustment to zero for every char
+        for char in target_seq:
+            char_complexity_adjustments[char] = 0
```

Comment:
Small issue, but it's better to initialize a float with `0.0`. If you do that, you also don't need the type annotation on the variables below, so instead of `t_factor: float = 0` you can just do `t_factor = 0.0` and the type checker will infer the correct type.